### DISCUSSION TITLE
[4.11] Nodes except bootstrap got "connection refused"

### DISCUSSION THREAD

**Discussion by Slonskiy** [okd-project/okd#1338] [Q&A]
After ignitions configs are generated, I'm trying to install and bootstrap cluster. With bootstrap node mainly is ok (sometimes I get this [issue](https://github.com/openshift/installer/issues/5858), but it fixes by removing etcd existing container).

But at master and probably worker (I don't try to provision worker nodes at now) got:
`GET error: Get “https://api-int.$cluster.$domain.com:22623/config/master”: dial tcp lookup api-int.$cluster.$domain.com on [::1]:53: read udp [::1]:$random_port->[::1=:53: read: connection refused`

At first, I need to setup local DNS for nodes VM's. But looks like it is not working.
For example, I install master node from live CD:
`./coreos-installer install --ignition-url http://api.$cluster.$domain.com:10808/master.ign --image-url https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/36.20220723.3.1/x86_64/fedora-coreos-36.20220723.3.1-metal.x86_64.raw.xz --network-dir /source/network/ --copy-network --ignition-hash sha512-$master_hash /dev/sda`

Address api.$cluster.$domain.com are resolved correctly because we on Live CD.
At /source/network/ folder only one file with follow contents:
```
[connection]
id=ens10
type=ethernet
autoconnect-priority=-999
interface-name=ens10
[ethernet]
[ipv4]
dns=10.0.0.2;
dns-search=$domain.com;
method=auto
[ipv6]
addr-gen-mode=stable-privacy
method=auto
[proxy]
```
Install finishing without any errors:
gpg: Signature made Mon 08 Aug 2022 08:32:25 PM CEST
gpg:                using RSA key 53DED2CB922D8B8D9E63FD18999F7CBF38AB71F4
gpg: Good signature from "Fedora (36) <fedora-36-primary@fedoraproject.org>" [ultimate]
Writing Ignition config
Copying networking configuration from /source/network/
Copying /source/network/ens10.nmconnection to installed system
Install complete.

But after reboot ignition cannot resolve address (but bootstrap node can do it without any problems, so it's not balancer or bootstrap node problem, how as it's not dns or network issue too).

The first idea was about StubListener. There is no settings for them at install time, and after install looks like resolving going via him ([::1]:53 in error). I was tried to turn it off by [this solution](https://github.com/okd-project/okd/blob/master/Guides/UPI/baremetal/manual-dns-and-fcos-dns-fix.md). But machine configs for nodes are applied after ignition gets them from bootstrap, but we can't resolve bootstrap address. I was tried append dns settings to ignition nodes configs, but there is now effect. Like follow (part of ignition that I manually append to generated one):
```
  "storage": {
    "files": [
      {
        "overwrite": true,
        "path": "/etc/systemd/resolved.conf.d/50-no-dns-stub.conf",
        "user": {
          "name": "root"
        },
        "contents": {
          "source": "data:text/plain;charset=utf-8;base64,W1Jlc29sdmVdCkROU1N0dWJMaXN0ZW5lcj1ubwo="
        },
        "mode": 420
      },
      {
        "overwrite": true,
        "path": "/etc/systemd/resolved.conf.d/75-static-dns-servers.conf",
        "user": {
          "name": "root"
        },
        "contents": {
          "source": "data:text/plain;charset=utf-8;base64,W1Jlc29sdmVdCkROUz0xMC4wLjAuMgpEb21haW5zPW15X2RvbWFpbi5jb20="
        },
        "mode": 420
      }
    ]
  }
```
But looks like at first at first ignition try to merge config from bootstrap, and only after that all other. 

So, if I understand situation right, network settings provided keyfile aren't work because StubListener used at first, StubListener settings aren't apply because they will apply only after bootstrap name will resolved to get master config from bootstrap. In this case I can't login to master node via ssh or console, so I don't know how to look current network settings. May be I do something wrong? Or this is a bug? I'm ready to provide any missing information if needed.

Version
FCOS 36.20220723.3.1
4.11.0-0.okd-2022-08-20-022919
UPI, bare metal

How reproducible
Issue is reproducing in all installations at specific environment.

**Comment by vrutkovs** [comment]
If its DHCP I'd expect DNS servers to pass it to the node. If its stable setup you may want to use NMState or `ip` kernel arg - https://man7.org/linux/man-pages/man7/dracut.cmdline.7.html

**Reply by Slonskiy** [reply]
DHCP can't pass correct dns settings in my case. DHCP pass correct IP and routes to the VM, but not dns servers and domain search string. It's because I can't setup DHCP by myself. To pass kernel args I need LOM or something like that. I think it's possible, will try. But I try to do automatic cluster deploy, so this way don't looks like good solution for me. Also will lurk about NMState, thank you.
Will try these variants and write here about results.

**Reply by Slonskiy** [reply]
How can I use NMState operator? Problem that I have with network settings is going before cluster node (master/worker) gets own config from bootstrap node. In this case I don't see that NMState can be helpful. 
As I see problem with applying custom dns settings to core os by openshift-installer. I don't think it's cluster settings problem.
And I'm not sure, but at one manual (don't have link for now) I was reading that ignition settings have more priority than dracut or afterburn.
